diamond.py

Sidduppal opened the pull request:
-  :art: Renames top_pct to bitscr_thresh which translates to bitscore threshold
-  :racehorse:  Uses all the available cpus as default 
-  :bug: Removes hardcoding of BLASTP
-  :art: Enables the searching of merged.dmp in case the taxid is not found in nodes.dmp
-  :art: :racehorse: gunzipped database opens with 'rt' mode
-  :memo:  Formats docstrings
- :bug: Fixes the code for `__sub__`

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+
+        Parameters
+        ----------
+        other_hit : Dict 
+            {
+            sseqid: {
+                "pident": float(pident),
+                "length": int(length),
+                "mismatch": int(mismatch),
+                "gapopen": int(gapopen),
+                "qstart": int(qstart),
+                "qend": int(qend),
+                "sstart": int(sstart),
+                "send": int(send),
+                "evalue": float(evalue),
+                "bitscore": float(bitscore),
+            }
+
+        Returns
+        -------
+        Dict
+            {qseqid: {DiamondResult}, ...}
```
Same comments as above. `other_hit` is not a `dict`, a `dict` is not returned and the parameter string is wrong

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+            {qseqid: {DiamondResult}, ...}
+        """
+        assert self == other_hit, f"qseqids do not match! {self} & {other_hit}"
+        self.sseqids.pop(list(other_hit.sseqids.keys())[0])
```
This is only removing the first `sseqid` from `self.sseqids`. This may also fail if the `other_hit.sseqids` are not in `self.sseqids`. You should be able to rework this so it will not raise a KeyError when trying to remove `other_hits.sseqids` that are not in the `self.sseqids`. This also needs to be fixed to follow the intended behavior

Sidduppal replied:
From my understanding of the script and trying with `pdb` I found that `hit.sseqids` will always have a single `sseqid`, and thus we only need to remove the first key. Please let me know in case I am missing something.

WiscEvan replied:
This may be true for when parsing the blast results, but this may not be true in other circumstances. Only removing one when instead the behavior intends to remove all, could result in a difficiult to find "_gotcha_" kind of 🐛

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+
+    Parameters
+    ----------
+    fasta : str
+        Path to fasta file whose database needs to be made
+        e.g. '<path/to/fasta/file>'
+    database : str
+        Path to the output  nr database file
+        e.g. '<path/to/database/file>'
```
These parameters are somewhat connected. You are referring to `nr` in the database parameter. But this function can handle _any_ fasta file and write out a respective diamond formatted database file from the provided `fasta`. I'm not sure `nr` should be mentioned here. The `database` parameter description should be more clear.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+    Returns
+    -------
+    str
+        Path to Diamond database
```
This should probably read "Path to diamond formatted database"

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+    Raises
+    ------
+    CalledProcessError
+        Failed to create Diamond database
```
```suggestion
        Failed to create diamond formatted database
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+
+    Raises
+    ------
+    CalledProcessError
```
Should this read `subprocess.CalledProcessError`?

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
     ----------
     fasta : str
-        </path/to/fasta/file>. May be amino acid or nucleotide sequences
+        Path to fasta file having the query sequence. May be amino acid or nucleotide sequences
```
The _fasta_ may contain multiple sequences. These would all be query sequences. The description suggests the fasta file should only contain one query sequence, which is false.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
     database : str
-        </path/to/diamond/formatted/database>.
+        Path to diamond formatted database
+        '</path/to/diamond/formatted/database>'
```
I'm not sure having both of these lines is helpful.

```suggestion

```

Sidduppal replied:
I understand that this is a biit redundant but I was thinking we should keep this to maintain consistency. Let me know what you think

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
-    verbose : boolean
-        log progress to terminal `verbose` (the default is False).
+        Path to output file
+        '</path/to/output/file>'
```
I'm not sure having both of these lines is helpful.

```suggestion

```

Sidduppal replied:
I understand that this is a biit redundant but I was thinking we should keep this to maintain consistency. Let me know what you think

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+        blastp to align protein query sequences against a protein reference database,
+        blastx to align translated DNA query sequences against a protein reference database, by default 'blastp'
+    evalue : float, optional
+        cutoff e-value to count hit as significant,, by default float('1e-5')
```
```suggestion
        cutoff e-value to count hit as significant, by default float('1e-5')
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
+    nproc : int, optional
+        Number of processors to be used, by default uses all the processors of the system
+    tmpdir : str, optional
+        Path to temporary directory
+        '</path/to/temporary/directory>' by default os.curdir
```
I'm not sure having both of these lines is helpful.

```suggestion
        Path to temporary directory, by default os.curdir
```

The default should probably _not_ be the current directory, but rather an actual temp directory. You could probably implement this using the [tempfile](https://docs.python.org/3/library/tempfile.html) module

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
-        `outfpath`
+        Path to BLAST results
 
     Raises
```
You need to update the docstring to reflect that a `subprocess.CalledProcessError` is raised.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -182,50 +337,60 @@ def blast(fasta, database, outfpath, blast_type='blastp', evalue=float('1e-5'),
-    top_pct : 0 < float <= 1
-        bitscore filter applied to each qseqid (the default is 0.9).
+    bitscr_thresh : 0 < float <= 1, optional
+        bitscore filter applied to each qseqid, by default 0.9
+        Used to determine whether the bitscore is above a threshold value. 
+        For example, if it is 0.9 then only bitscores >= 0.9 * the top bitscore are accepted
```
This would be applied to the sseqids. This could probably be more concise/clear

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -283,18 +452,28 @@ def parse(results, top_pct=0.9, verbose=False):
                 continue
-            if bitscore >= top_pct * topbitscore:
+            if bitscore >= bitscr_thresh * topbitscore:
+                import pdb
+
+                pdb.set_trace()
```
`pdb` should not be in the code
```suggestion
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,78 +518,104 @@ def add_taxids(hits, database, verbose=True):
-            #     raise DatabasesOutOfDateError(f'{sseqid} deprecated/suppressed/removed')
-            hit.sseqids[sseqid].update({'taxid':taxid})
+            if not taxid:
+                taxid = ncbi.merged.get(taxid, taxid)
```
If the taxid is `None`, how are you searching for this in merged? This appears to be a bug

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,78 +518,104 @@ def add_taxids(hits, database, verbose=True):
-    desc = f'Translating {n_qseqids} qseqids\' accessions to taxids'
-    for qseqid,hit in tqdm(hits.items(), disable=disable, total=n_qseqids, desc=desc, leave=False):
+    desc = f"Translating {n_qseqids} qseqids' accessions to taxids"
+    ncbi = NCBI(NCBI_DIR, verbose)  # Instantiating the NCBI class
```
Is this necessary at this stage?

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,78 +518,104 @@ def add_taxids(hits, database, verbose=True):
+        "--nproc", help="number of processors to use", default=mp.cpu_count(), type=int
+    )
+    parser.add_argument("--tmpdir", help="</path/to/tmp/directory>", default=os.curdir)
+    parser.add_argument(
+        "--bitscr-thresh",
+        help="threshold above which bitscores will be selected",
+        default=0.9,
+    )
```
This is an end-user argument and should be more descriptive. e.g. `--bitscore-threshold`. The help text is not very informative

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,78 +518,104 @@ def add_taxids(hits, database, verbose=True):
+        type=int,
+    )
+    parser.add_argument(
+        "--nproc", help="number of processors to use", default=mp.cpu_count(), type=int
```
Do you think `nproc` or `cpus` is more clear to the end user?

Sidduppal replied:
`cpus` would be better. Do you thin we should change this in other scripts as well?

WiscEvan replied:
I think probably only necessary when it is front facing. But kind of a judgement call. Which do you prefer? Either works for me.

Sidduppal replied:
I think `cpus` makes it clear to the end-user as well as the developer. 
We can definitely have a different front-facing variable but it may cause some minor confusions in the future.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
 
 logger = logging.getLogger(__name__)
 
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+NCBI_DIR = os.path.join(BASE_DIR, "databases", "ncbi")
```
I'm not sure this is necessary.

Sidduppal replied:
I added these to used `self.merged` here:
https://github.com/Sidduppal/Autometa/blob/72fc07b3dd0d3503f71da7447b04aa653b0c85e6/autometa/common/external/diamond.py#L551-L552

But after this [comment](https://github.com/KwanLab/Autometa/pull/87#discussion_r436074791), it looks like my solution was not correct after all. I will re-think how to search the `merge.dmp`. any suggestions?

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -29,146 +29,301 @@
-from autometa.common.utilities import make_pickle,file_length
+from autometa.common.utilities import make_pickle, file_length
 from autometa.common.external import prodigal
+from autometa.taxonomy.ncbi import NCBI
```
I do not think this is necessary

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -182,50 +337,60 @@ def blast(fasta, database, outfpath, blast_type='blastp', evalue=float('1e-5'),
-def parse(results, top_pct=0.9, verbose=False):
-    """Retrieve diamond results from output table
+
+def parse(results, bitscr_thresh=0.9, verbose=False):
```
I'm not sure `bitscr_thresh` is clear as a variable name. When inspecting blast results, the bitscore is usually a float that can range to the 1000s. However, we are asking for a float between 0 and 1. I think this is confusing.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -283,18 +452,28 @@ def parse(results, top_pct=0.9, verbose=False):
 def add_taxids(hits, database, verbose=True):
-    """Translates accessions to taxid translations from prot.accession2taxid.gz
+    """
+    Translates accessions to taxid translations from prot.accession2taxid.gz
```
This translates accessions to taxids. Not taxid translations

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -283,18 +452,28 @@ def parse(results, top_pct=0.9, verbose=False):
-    """Translates accessions to taxid translations from prot.accession2taxid.gz
+    """
+    Translates accessions to taxid translations from prot.accession2taxid.gz
+    If an accession number is no longer available (either due to being suppressed, 
+    deprecated or removed by NCBI), then merged.dmp is searched. If not found even there
+    then None is returned.
```
If you can not retrieve a taxid, how are you able to search merged?

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -582,12 +555,14 @@ def main():
     parser.add_argument(
-        "--bitscr-thresh",
-        help="threshold above which bitscores will be selected",
+        "--top_percentile",
+        help="pecentile above which hits will be retreived",
```
```suggestion
        "--top-percentile",
        help="percentile above which hits will be retrieved",
```

Sidduppal replied:
Why not use underscores (`_`)?

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -582,12 +555,14 @@ def main():
-        "--nproc", help="number of processors to use", default=mp.cpu_count(), type=int
+        "--cpus", help="number of processors to use", default=mp.cpu_count(), type=int
+    )
+    parser.add_argument(
+        "--tmpdir", help="</path/to/tmp/directory>", default=tempfile.mkdtemp()
```
Depending on how you handle the comments about the `tmpdir` above, you may be able to remove this.

WiscEvan commented:
It appears the docs are failing from the `tempdir` argument. The default value you are setting as `tempfile.mkdtemp` when `tempfile is not imported. This should be resolved if this parameter ends up being removed. See https://github.com/KwanLab/Autometa/pull/87#discussion_r439560687

Sidduppal commented on autometa/common/external/diamond.py:
```diff
@@ -330,6 +307,8 @@ def blast(
+        Failed to run blast
     """
     if not os.path.exists(fasta):
         raise FileNotFoundError(fasta)
```
1. I think `if verbose` can be a handy tool if someone is running the script as a standalone module. What do you think about it? 
Else I can remove the `if verbose` statements and just use the logging statement for all the functions.

Sidduppal replied:
I think we should keep `tmpdir` as an input from the user. It may be the case that the user needs to specify a different directory rather than the temporary directory like in the case of space issues. I don't think we should use a `tempdir` flag with 1diamond` if we are randomly creating a temporary directory, as `diamond` itself creates a temporary directory to use. 
From what I read from the diamond documentation the `tempdir` flag is only needed in cases where the output directory does not has enough space available.
Let me know what you think about it

Sidduppal commented:
I was reading about the `assert` method. It will raise an error in case the query sequences are not equal. Do we want an error to be raised? If yes, should be mention that in the docstrings?

WiscEvan commented:
> I was reading about the `assert` method. It will raise an error in case the query sequences are not equal. Do we want an error to be raised? If yes, should be mention that in the docstrings?

You make a good point. What do you think? I think arguments could be made for either implementation.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -28,204 +28,347 @@
+        proc.check_returncode()
+    except subprocess.CalledProcessError as err:
+        raise err
+    shutil.rmtree(tmpdir, ignore_errors=True)
```
Have you tested this? What happens if the user were to specify the current working directory as their tmpdir? What about the autometa output directory? Do we want to remove everything in this directory?

WiscEvan commented on docs/parse_argparse.py:
```diff
@@ -57,7 +57,7 @@ def get_argparse_block(fpath):
     """
     writing = False
-    outlines = "import os\nimport multiprocessing as mp\n"
+    outlines = "import os\nimport multiprocessing as mp\nimport tempfile\n"
```
This seems kind of like a heavy handed approach for one specific file. I'm not sure this needs to be included in every script.

WiscEvan commented on docs/parse_argparse.py:
```diff
@@ -105,30 +103,38 @@ def get_usage(argparse_lines):
+
+    Raises
+    ------
+    CalledProcessError
```
```suggestion
    subprocess.CalledProcessError
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,35 +492,82 @@ def add_taxids(hits, database, verbose=True):
+        "--cpus", help="number of processors to use", default=mp.cpu_count(), type=int
+    )
+    parser.add_argument(
+        "--tmpdir", help="</path/to/tmp/directory>", default=tempfile.mkdtemp()
```
Is placing a default value necessary here? The function (`blastp`) already places a default value for `tmpdir`. This seems like it is adding more headache and not really doing anything for the end user.

Sidduppal commented:
> 

> > I was reading about the `assert` method. It will raise an error in case the query sequences are not equal. Do we want an error to be raised? If yes, should be mention that in the docstrings?
> 
> You make a good point. What do you think? I think arguments could be made for either implementation.

We should keep `assert` and raise an error if the condition is False. However, that should also be mentioned in the docstring. From my understanding, the error would not be raised if the script is being used as the part of the pipeline as the following lines (that are above `hits[hit.qseqid] += hit`) make sure that the query sequences are the same ones.

https://github.com/KwanLab/Autometa/blob/ddc2bf951d6c4a8c29e9e2301e15a762327b0267/autometa/common/external/diamond.py#L287-L291

Raising an error would be good when running the script as a stand-alone module or for future developers to test the module.
Let me know what you think about this.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -28,204 +28,359 @@
+        assert (
+            self == other_hit
+        ), f"qseqids do not match! {self.qseqid} & {other_hit.qseqid}"
+        for key in list(other_hit.sseqids.keys()):
+            try:
+                self.sseqids.pop(key)
+            except KeyError as err:
+                raise err
```
A few problems here.

1. There is no need to use the `list` function when also employing the `dict.keys` method. When iterating through a dictionary, the default behavior is to iterate through the keys. So the method and the function are unnecessary.
2. It would be more helpful to be descriptive with your variables.  i.e. `sseqid` rather than `key`.
2. The error message is not helpful and this is 3 extra lines to give the same exact message as just brute forcing `self.sseqids.pop(key)`
3. I think the `try` and `except` block could be useful here, but you need to provide more helpful information.

```suggestion
        for sseqid in other_hit.sseqids:
            try:
                self.sseqids.pop(sseqid)
            except KeyError:
                raise KeyError(<'Some useful error message here'>)
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -28,204 +28,359 @@
+    subprocess.CalledProcessError
+        Failed to create diamond formatted database
+    """
+    cmd = f"diamond makedb --in {fasta} --db {database} -p {cpus}"
+    logger.debug(f"{cmd}")
+    proc = subprocess.run(
+        cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True
+    )
+    try:
+        proc.check_returncode()
+    except subprocess.CalledProcessError as err:
+        raise err
```
1. Since you returning the same error message that is being returned from `subprocess.CalledProcessError`, you could place this within `subprocess.run(check=True)`
2. You are running `subprocess.run` with `shell=True`, this has been [documented as being a security risk](https://docs.python.org/3/library/subprocess.html#security-considerations). So we should try to minimize this as much as possible.
```suggestion
    cmd = ["diamond", "makedb", "--in", fasta, "--db", database ,"-p", cpus]
    subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
    logger.debug(" ".join(cmd))
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -28,204 +28,359 @@
-        proc = subprocess.run(cmd, stdout=stdout, stderr=stderr)
-    if proc.returncode:
-        raise OSError(f'DiamondFailed:\nArgs:{proc.args}\nReturnCode:{proc.returncode}')
+    proc = subprocess.run(
+        cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=False
+    )
+    try:
+        proc.check_returncode()
+    except subprocess.CalledProcessError as err:
+        raise err
```
1. Similar comments as in the `makedatabase` function.
2. `shell=False` is the default and not necessary

```suggestion
    subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -343,35 +504,80 @@ def add_taxids(hits, database, verbose=True):
+    import argparse
+    import os
+
+    parser = argparse.ArgumentParser(
+        description="""
+    Retrieves blastp hits with provided input assembly
+    """
+    )
```
Since you are specifying multiple defaults in the argparse parser. You should apply the defaults formatter.

https://github.com/KwanLab/Autometa/blob/ddc2bf951d6c4a8c29e9e2301e15a762327b0267/docs/template.py#L89-L94

```suggestion
    parser = argparse.ArgumentParser(
        description="Retrieves blastp hits with provided input assembly",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
```

Sidduppal commented:
I think I made a mistake in the last commit by removing the default value of `tempdir`. Doing this results in `tmpdir=args.tmpdir` being equal to `None` which is an invalid input for diamond blast. From how I look at it we have three solutions:
1. Set the default value as `os.curdir`, so we don't have to add anything more in `parse.argparse.py`
2. Hardcode `import tempfile` in `parse.argaprse.py`
3. Write `import tempfile` just below `import argparse` in `diamond.py`
I was more leaning towards the third option as it is more versatile, but I would like to hear your views on it as well.

WiscEvan commented:
> I think I made a mistake in the last commit by removing the default value of `tempdir`. Doing this results in `tmpdir=args.tmpdir` being equal to `None` which is an invalid input for diamond blast. From how I look at it we have three solutions:
> 
> 1. Set the default value as `os.curdir`, so we don't have to add anything more in `parse.argparse.py`
> 2. Hardcode `import tempfile` in `parse.argaprse.py`
> 3. Write `import tempfile` just below `import argparse` in `diamond.py`
>    I was more leaning towards the third option as it is more versatile, but I would like to hear your views on it as well.

Why not just not pass in `args.tmpdir` if it is not specified? This seems like the simplest solution

Sidduppal commented:
> > I think I made a mistake in the last commit by removing the default value of `tempdir`. Doing this results in `tmpdir=args.tmpdir` being equal to `None` which is an invalid input for diamond blast. From how I look at it we have three solutions:
> > 
> > 1. Set the default value as `os.curdir`, so we don't have to add anything more in `parse.argparse.py`
> > 2. Hardcode `import tempfile` in `parse.argaprse.py`
> > 3. Write `import tempfile` just below `import argparse` in `diamond.py`
> >    I was more leaning towards the third option as it is more versatile, but I would like to hear your views on it as well.
> 
> Why not just _not_ pass in `args.tmpdir` if it is not specified? This seems like the simplest solution

Do you mean something like:
```python
tempdir=tempfile.mkdtemp() if args.tempdir==None else args.tempdir 
```

WiscEvan commented:
You need to update this PR with the most recent changes from `dev`

Sidduppal commented:
Yes, but I didn't push/commit anything. I was thinking of resolving the issue of tempdir and then pushing. Any ideas on which option to go with?

WiscEvan commented:
> > > I think I made a mistake in the last commit by removing the default value of `tempdir`. Doing this results in `tmpdir=args.tmpdir` being equal to `None` which is an invalid input for diamond blast. From how I look at it we have three solutions:
> > > 
> > > 1. Set the default value as `os.curdir`, so we don't have to add anything more in `parse.argparse.py`
> > > 2. Hardcode `import tempfile` in `parse.argaprse.py`
> > > 3. Write `import tempfile` just below `import argparse` in `diamond.py`
> > >    I was more leaning towards the third option as it is more versatile, but I would like to hear your views on it as well.
> > 
> > 
> > Why not just _not_ pass in `args.tmpdir` if it is not specified? This seems like the simplest solution
> 
> Do you mean something like:
> `tmpdir=not args.tmpdir if args.tmpdir == None else args.tmpdir`
> This is also returning error. It needs a directory as input and not a boolean

`tmpdir=not args.tmpdir if args.tmpdir == None else args.tmpdir` is too verbose and will result in a boolean for one of the values.

More along the lines of what you are passing in to your function.

I was thinking of when you pass `args.tmpdir` into the function:

1. 
```python

tmpdir = tempfile.mkdtemp() if not args.tmpdir else args.tmpdir
result = blast(
        fasta=args.fasta,
        database=args.database,
        outfpath=args.outfile,
        blast_type=args.blast_type,
        evalue=args.evalue,
        maxtargetseqs=args.maxtargetseqs,
        cpus=args.cpus,
        tmpdir=tmpdir,
        force=args.force,
        verbose=args.verbose)
```

2. 
```python
if not args.tmpdir:
    result = blast(
        fasta=args.fasta,
        database=args.database,
        outfpath=args.outfile,
        blast_type=args.blast_type,
        evalue=args.evalue,
        maxtargetseqs=args.maxtargetseqs,
        cpus=args.cpus,
        force=args.force,
        verbose=args.verbose)
else:
    result = blast(
        fasta=args.fasta,
        database=args.database,
        outfpath=args.outfile,
        blast_type=args.blast_type,
        evalue=args.evalue,
        maxtargetseqs=args.maxtargetseqs,
        cpus=args.cpus,
        tmpdir=args.tmpdir,
        force=args.force,
        verbose=args.verbose)
```

Both of these routines seem round about though. Is having `tempfile.mkdtemp()` as a default necessary?

From the [diamond manual](http://gensoft.pasteur.fr/docs/diamond/0.8.29/diamond_manual.pdf):

>3.6 Memory & performance options
>
>--tmpdir/-t <directory>
Directory to be used for temporary storage. This is set to the output directory by default.
The amount of disk space that will be used depends on the program’s settings and your
data. As a general rule you should ensure that 100 GB of disk space are available here.
If you run the program in a cluster environment, and disk space is only available over a
slow network based file system, you may want to set the --tmpdir option to /dev/shm. This
will keep temporary information in memory and thus increase the program’s memory usage
substantially.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -28,204 +28,359 @@
+        "--db",
         database,
-        '--evalue',
+        "--evalue",
         evalue,
-        '--max-target-seqs',
+        "--max-target-seqs",
         maxtargetseqs,
-        '--threads',
+        "--threads",
         cpus,
-        '--outfmt',
-        '6',
-        '--out',
+        "--outfmt",
+        "6",
+        "--out",
         outfpath,
-        '--tmpdir',
-        tmpdir]
+        "--tmpdir",
+        tmpdir,
+    ]
```
It seems to me, the only time you will need to add `--tmpdir tmpdir` to the command is when this is supplied by the user. Otherwise, I think diamond will handle this. Similarly, you could place the logic _within_ this function if you would like to set the tmpdir using the `tempfile.mkdtemp` method. The default `tmpdir` parameter could be `None`. If the default were to be `None`, then you could assign the `tmpdir` variable to `tempfile.mkdtemp()`.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -157,66 +269,72 @@ def blast(
-        if not empty:
-            if verbose:
-                logger.warning(f"FileExistsError: {outfpath}. To overwrite use --force")
+        if os.path.getsize(outfpath) and verbose:
+            logger.warning(f"FileExistsError: {outfpath}. To overwrite use --force")
```
Why is returning here dependent on the user specifying verbose?

```suggestion
        if os.path.getsize(outfpath):
            logger.warning(f"FileExistsError: {outfpath}. To overwrite use --force")
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -157,66 +269,72 @@ def blast(
+        tmpdir = tempfile.mkdtemp()
     if verbose:
-        logger.debug(f"Diamond{blast_type.title()} {fasta} against {database}")
+        logger.debug(f"Diamond{blast_type} {fasta} against {database}")
```
minor changes for whitespace and executable

```suggestion
        logger.debug(f"diamond {blast_type} {fasta} against {database}")
```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -405,43 +527,19 @@ def add_taxids(hits, database, verbose=True):
-if __name__ == "__main__":
-    import argparse
+def main():
     import os
```
This appears to be imported twice?

```suggestion

```

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -157,66 +269,72 @@ def blast(
     blast_type = blast_type.lower()
     if blast_type not in ["blastp", "blastx"]:
         raise ValueError(f"blast_type must be blastp or blastx. Given: {blast_type}")
+    if tmpdir == None:
+        tmpdir = tempfile.mkdtemp()
```
Typically if you are checking if `something` is `None` you can just write `if something`.
You could simplify this to: `tmpdir = tmpdir if tmpdir else tempfile.mkdtemp()`.

That being said, this overwrites the option for diamond to run its default tmpdir. I think if `tmpdir` is `None` then the command should not be added from the list of commands or could be removed from the list of commands.

```python
    cmd = [...]
    if tmpdir:
        cmd.extend(["--tmpdir", tmpdir])
```

Sidduppal commented:
I have addressed the comment regarding using the default temporary directory (specified by diamond) when the user does not specify anything. However, I was thinking this will only work when the script is being run as a standalone module. When the script will be run as the part of the pipe, the parameter, `tmpdir=tempfile.mkdtemp()` in the function definition would create a temporary directory. I don't think this is a major issue but I just wanted to bring your attention to it.

WiscEvan commented on autometa/common/external/diamond.py:
```diff
@@ -462,14 +558,43 @@ def main(args):
-        "--top-pct", help="top percentage of hits to retrieve", default=0.9
+        "--cpus", help="number of processors to use", default=mp.cpu_count(), type=int
+    )
+    parser.add_argument("--tmpdir", help="</path/to/tmp/directory>")
```
The help text for this could be more descriptive. It is unclear why specifying a tmp directory is necessary. What is using the tmp directory?

Something like "Path to tmp directory used by diamond (See diamond manual section 3.6 for details)."

Sidduppal commented on autometa/common/external/diamond.py:
```diff
@@ -157,66 +269,71 @@ def blast(
     blast_type="blastp",
     evalue=float("1e-5"),
     maxtargetseqs=200,
-    cpus=0,
-    tmpdir=os.curdir,
+    cpus=mp.cpu_count(),
+    tmpdir=tempfile.mkdtemp(),
     force=False,
     verbose=False,
 ):
```
You are right, the default doesn't have to be `tempfile.mkdtemp()`. Looks like I made a silly mistake by not altering it already :facepalm: The default is now `None` thanks for correcting it :+1: :clinking_glasses:

Sidduppal commented on autometa/common/external/diamond.py:
```diff
@@ -108,29 +116,110 @@ def __init__(
+        assert (
+            self == other_hit
+        ), f"qseqids do not match! {self.qseqid} & {other_hit.qseqid}"
+        for sseqid in other_hit.sseqids.keys():
+            try:
+                self.sseqids.pop(sseqid)
+            except KeyError:
+                raise KeyError(
+                    "The query sequence corresponsing to the DiamondResult is not present!"
+                )
```
Thanks for the link I was not aware of that :beers: :bowing_man: :checkered_flag:

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -231,28 +347,33 @@ def blast(
 
-def parse(results, top_pct=0.9, verbose=False):
-    """Retrieve diamond results from output table
+def parse(results, top_percentile=0.9, verbose=False):
```
You ignored my request in issue #36. Please read it again:

>In the script, there is an argument/variable called top_pct or a variant of this name. This is a float value between 0 and 1 (although it can't exactly be zero), used to determine if the bitscore of hits is above a threshold. For example, if it is 0.9 then only bitscores >= 0.9 * the top bitscore are accepted. The problem here is not one of programmatic logic, but rather readability for contributors. top_pct looks like "top percentage", which would suggest, for example, that for a value of 0.9, 90% will be kept. However, the opposite is true in this case - only 10% will be kept. This is compounded by the help message in the if __name__ == '__main__': block, which describes this argument as "top percentage of hits to retrieve". So what I am saying is that the variable name and documentation is confusing. Perhaps it could be called "bitscore_tolerance" or something? Accompanied by a change in documentation and inclusion of explanatory comments.

The variable does not describe a "top percentile" to keep, but rather some sort of threshold.

Sidduppal replied:
I was thinking of`bitscr_filter` with the corresponding `help` text:
"hits with bitscore greater than bitscr-filter*top bitscore will be kept".
Other options that I thought were `bitscore_threshold` and the one you suggested i.e. `bitscr_tolerance`. I was moe inclined towards `bitscr_filter`.
What do you think about it? Any ideas?

jason-c-kwan replied:
`bitscore_filter` is OK, but can you not abbreviate? If you have a decent text editor it should autocomplete anyway.

Sidduppal replied:
Alright!
sure I can write the full name. I don't know why I was of the view that developers prefer variable names that have less character (i.e. abbreviated), seems like I was wrong. I'll make the change and push the changes!

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -347,7 +472,10 @@ def add_taxids(hits, database, verbose=True):
     database : str
-        </path/to/prot.accession2taxid.gz>
+        Path to prot.accession2taxid.gz database
+        e.g. </path/to/prot.accession2taxid.gz>
```
This line is not needed.

Sidduppal replied:
I am not sure which line(s) are you pointing towards.
Also, I kept both the verbal description as well as the example to maintain consistency throughout the codebase.
Let me know if you only need the description, I would the remove the "e.g." from other function like `blast` as well

jason-c-kwan replied:
Line 476. It doesn't add any information that is not in line 475, especially because you specify the exact filename in line 475.

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -462,14 +557,46 @@ def main(args):
+        help="Path to directory which will be used for temporary storage by diamond",
+    )
+    parser.add_argument(
+        "--top-percentile",
+        help="percentile above which hits will be retrieved",
+        default=0.9,
```
See other comment

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -42,12 +42,18 @@
+    Allows for better handling of subject sequence IDs that hit to a qseqid.
+
+    Includes methods used to modify the DiamondResult object (add or remove sseqid from sseqids dictionary),
+    check if two DiamondResult object have the same qseqid and return a user firendly and unambiguous output
```
Grammatical and spelling error:

```suggestion
    check if two DiamondResult objects have the same qseqid and return a user friendly and unambiguous output
```

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -42,12 +42,18 @@
+
+    Includes methods used to modify the DiamondResult object (add or remove sseqid from sseqids dictionary),
+    check if two DiamondResult object have the same qseqid and return a user firendly and unambiguous output
+    from str() ad repr() respectively. Also used to returns the sseqid with the highest bitscore amongst all
+    the subject sequences that hit a query. These methds come in handy when retrieving diamond results from output table.
```
More spelling errors:

```suggestion
    from str() ad repr() respectively. Also used to return the sseqid with the highest bitscore amongst all
    the subject sequences that hit a query. These methods come in handy when retrieving diamond results from output table.
```

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -297,8 +301,7 @@ def blast(
     tmpdir : str, optional
-        Path to temporary directory
-        '</path/to/temporary/directory>' by default output directory
+        Path to temporary directory by default, output directory
```
Should be two sentences.

```suggestion
        Path to temporary directory. By default, same as the output directory
```

Sidduppal replied:
I was using VS code extension for docstring with the type set to `numpy`. Going through `numpy` docstring documentation I was not able to find a standard way to mention the default value, but since we all are now using VS code with the same extension, I think having "by default" in the same sentence would help maintain consistency. What do you think about this?

jason-c-kwan replied:
No. I put this suggestion because the sentence structure you put is wrong. "Path to temporary directory by default" means that by default it is a temporary directory, and I suppose if the user chooses it will be some other type of directory? It doesn't make much sense. Then there is the other part of the sentence. "Path to temporary directory by default, output directory". The "output directory" part is out there on its own as if it was the start of the next clause, but then there is no verb or anything else. The punctuation I suggested makes it unambiguous that "By default" belongs with "output directory". I think your priority should be clarity before consistency. (But also I highly doubt that the numpy docstrings are grammatically wrong).

jason-c-kwan replied:
Please address this before we merge the PR.

jason-c-kwan commented on autometa/common/external/diamond.py:
```diff
@@ -46,9 +46,9 @@ class DiamondResult:
-    from str() ad repr() respectively. Also used to returns the sseqid with the highest bitscore amongst all
-    the subject sequences that hit a query. These methds come in handy when retrieving diamond results from output table.
+    check if two DiamondResult objects have the same qseqid and return a user friendly and unambiguous output
+    from str() ad repr() respectively. Also used to return the sseqid with the highest bitscore amongst all
```
Found another typo:
```suggestion
    from str() and repr() respectively. Also used to return the sseqid with the highest bitscore amongst all
```
